File: python-sdk/hsafa/resources/runs.py
from typing import Any, Dict, Optional

from ..http import HttpClient
from ..sse import SSEStream


class RunsResource:
    def __init__(self, http: HttpClient):
        self._http = http

    def list(
        self,
        space_id: Optional[str] = None,
        agent_entity_id: Optional[str] = None,
        agent_id: Optional[str] = None,
        status: Optional[str] = None,
        limit: Optional[int] = None,
        offset: Optional[int] = None,
    ) -> Dict[str, Any]:
        return self._http.get("/api/runs", {
            "smartSpaceId": space_id,
            "agentEntityId": agent_entity_id,
            "agentId": agent_id,
            "status": status,
            "limit": limit,
            "offset": offset,
        })

    def get(self, run_id: str) -> Dict[str, Any]:
        return self._http.get(f"/api/runs/{run_id}")

    def create(
        self,
        agent_entity_id: str,
        agent_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        return self._http.post("/api/runs", {
            "agentEntityId": agent_entity_id,
            "agentId": agent_id,
            "metadata": metadata,
        })

    def cancel(self, run_id: str) -> Dict[str, Any]:
        return self._http.post(f"/api/runs/{run_id}/cancel")

    def delete(self, run_id: str) -> Dict[str, Any]:
        return self._http.delete(f"/api/runs/{run_id}")

    def get_events(self, run_id: str) -> Dict[str, Any]:
        return self._http.get(f"/api/runs/{run_id}/events")

    def subscribe(self, run_id: str, since: Optional[str] = None) -> SSEStream:
        params: Dict[str, Any] = {}
        if since:
            params["since"] = since
        url = self._http.build_url(f"/api/runs/{run_id}/stream", params)
        return SSEStream(url, self._http.get_auth_headers(), reconnect=True).start()

File: hsafa-core/old-hsafa/node-sdk/src/resources/runs.ts
import type { HttpClient } from '../http.js';
import type {
  Run,
  RunEvent,
  HsafaStream,
  CreateRunParams,
  ListRunsParams,
} from '../types.js';
import { SSEStream } from '../sse.js';

export class RunsResource {
  constructor(private http: HttpClient) {}

  async list(params?: ListRunsParams): Promise<{ runs: Run[] }> {
    return this.http.get('/api/runs', {
      smartSpaceId: params?.smartSpaceId,
      agentEntityId: params?.agentEntityId,
      agentId: params?.agentId,
      status: params?.status,
      limit: params?.limit,
      offset: params?.offset,
    });
  }

  async get(runId: string): Promise<{ run: Run }> {
    return this.http.get(`/api/runs/${runId}`);
  }

  async create(params: CreateRunParams): Promise<{ run: Run }> {
    return this.http.post('/api/runs', params);
  }

  async cancel(runId: string): Promise<{ success: boolean; status: string }> {
    return this.http.post(`/api/runs/${runId}/cancel`);
  }

  async delete(runId: string): Promise<{ success: boolean }> {
    return this.http.delete(`/api/runs/${runId}`);
  }

  async getEvents(runId: string): Promise<{ events: RunEvent[] }> {
    return this.http.get(`/api/runs/${runId}/events`);
  }

  subscribe(runId: string, options?: { since?: string }): HsafaStream {
    const baseUrl = this.http.getBaseUrl();
    const headers = this.http.getAuthHeaders();
    const params = new URLSearchParams();

    if (options?.since) {
      params.set('since', options.since);
    }

    const queryStr = params.toString();
    const url = `${baseUrl}/api/runs/${runId}/stream${queryStr ? `?${queryStr}` : ''}`;

    return new SSEStream({
      url,
      headers,
      reconnect: true,
    });
  }
}
